src/EditorFeatures/Core/Implementation/DocumentationComments/AbstractDocumentationCommentSnippetService.cs

```diff
@@ -0,0 +1,411 @@
+            {
+                return null;
+            }
+
+            if (!IsMemberDeclaration(targetMember))
+            {
+                return null;
+            }
+
+            if (targetMember.SpanStart < documentationComment.SpanStart)
+            {
+                return null;
+            }
+
+            return targetMember;
+        }
+
+        private static void AddLineBreaks(IList<string> lines, string newLine)
+        {
+            for (var i = 0; i < lines.Count; i++)
+            {
+                lines[i] = lines[i] + newLine;
+            }
+        }
+
+        public DocumentationCommentSnippet? GetDocumentationCommentSnippetOnEnterTyped(SyntaxTree syntaxTree, SourceText text, int position, DocumentOptionSet options, CancellationToken cancellationToken)
+        {
+            // Don't attempt to generate a new XML doc comment on ENTER if the option to auto-generate
+            // them isn't set. Regardless of the option, we should generate exterior trivia (i.e. /// or ''')
+            // on ENTER inside an existing XML doc comment.
+
+            if (options.GetOption(FeatureOnOffOptions.AutoXmlDocCommentGeneration))
+            {
+                var result = GenerateDocumentationCommentAfterEnter(syntaxTree, text, position, options, cancellationToken);
+                if (result != null)
+                {
+                    return result;
+                }
+            }
+
+            return GenerateExteriorTriviaAfterEnter(syntaxTree, text, position, options, cancellationToken);
+        }
+
+        private DocumentationCommentSnippet? GenerateDocumentationCommentAfterEnter(SyntaxTree syntaxTree, SourceText text, int position, DocumentOptionSet options, CancellationToken cancellationToken)
+        {
+            // Find the documentation comment before the new line that was just pressed
+            var token = GetTokenToLeft(syntaxTree, position, cancellationToken);
+            if (!IsDocCommentNewLine(token))
+            {
+                return null;
+            }
+
+            var documentationComment = token.GetAncestor<TDocumentationComment>();
+            if (documentationComment == null)
+            {
+                return null;
+            }
+
+            if (!IsSingleExteriorTrivia(documentationComment))
+            {
+                return null;
+            }
+
+            var targetMember = GetTargetMember(documentationComment);
+            if (targetMember == null)
+            {
+                return null;
+            }
+
+            // Ensure that the target member is only preceded by a single documentation comment (our ///).
+            if (GetPrecedingDocumentationCommentCount(targetMember) != 1)
+            {
+                return null;
+            }
+
+            var line = text.Lines.GetLineFromPosition(documentationComment.FullSpan.Start);
+            if (line.IsEmptyOrWhitespace())
+            {
+                return null;
```

Comment:
I don't love all of these return nulls, but TryGet felt wierd because cancellationToken wouldn't be the last parameter, having a Success property on the result seemed silly. Open to suggestions here.